extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsExtractor.java

```diff
@@ -45,6 +45,7 @@
      */
     @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
     private Optional<Boolean> optCommentsDisabled = Optional.empty();
+    private JsonObject ajaxJson;
```

Comment:
No documentation.
I'm also not sure if it's good to keep this in the memory, because it might be rather big.

I think it would be better when you instantly fetch the comment count when the extractor is called (like the optCommentsDisabled).
This also hides the parameter inside ``getNextPage`` below.

Reply:
> I'm also not sure if it's good to keep this in the memory, because it might be rather big.

It's a JSON object, and these classes shouldn't be reused, right?

> I think it would be better when you instantly fetch the comment count when the extractor is called (like the optCommentsDisabled).

Wouldn't that result in 2 API calls? That's exactly what I want to avoid.

Reply:
> It's a JSON object

Not sure how that changes the situation

> and these classes shouldn't be reused

The class itself is not reused but the comments extractor is e.g. hold in memory by https://github.com/TeamNewPipe/NewPipeExtractor/blob/dev/extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsInfo.java#L73


> Wouldn't that result in 2 API calls? That's exactly what I want to avoid.

No, ``optCommentsDisabled`` also does exactly one API call.

Reply:
> No, optCommentsDisabled also does exactly one API call.

That's because it's extracted in findInitialCommentsToken(), which can be called without any stateless parameters, unlike here where we require a continuation token.

I don't what's the problem since we do exactly this in the `YoutubeStreamExtractor` class: https://github.com/TeamNewPipe/NewPipeExtractor/blob/65129e65a81563c81cd8dbcb6c64e7570bd2910c/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java#L93-L101

I had a look at implementing your suggestion, but I couldn't find a simple way, which wouldn't create 2 requests.

Reply:
How I would have implemented it:

<details><summary>CommentsExtractor.java</summary>

```diff
--- a/extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsExtractor.java
+++ b/extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsExtractor.java
@@ -6,9 +6,12 @@ import org.schabi.newpipe.extractor.exceptions.ExtractionException;
 import org.schabi.newpipe.extractor.exceptions.ParsingException;
 import org.schabi.newpipe.extractor.linkhandler.ListLinkHandler;
 
+import java.util.OptionalInt;
+
 import javax.annotation.Nonnull;
 
 public abstract class CommentsExtractor extends ListExtractor<CommentsInfoItem> {
+    public static int NO_COMMENT_COUNT = -1;
 
     public CommentsExtractor(final StreamingService service, final ListLinkHandler uiHandler) {
         super(service, uiHandler);
@@ -22,6 +25,14 @@ public abstract class CommentsExtractor extends ListExtractor<CommentsInfoItem>
         return false;
     }
 
+    /**
+     *
+     * @return total number of comments or {@link #NO_COMMENT_COUNT} if no count was extracted.
+     */
+    public int getCommentsCount() {
+        return NO_COMMENT_COUNT;
+    }
+
     @Nonnull
     @Override
     public String getName() throws ParsingException {
```

</details>

<details><summary>YoutubeCommentsExtractor.java</summary>

```diff
--- a/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsExtractor.java
+++ b/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsExtractor.java
@@ -1,7 +1,6 @@
 package org.schabi.newpipe.extractor.services.youtube.extractors;
 
-import static org.schabi.newpipe.extractor.services.youtube.YoutubeParsingHelper.getJsonPostResponse;
-import static org.schabi.newpipe.extractor.services.youtube.YoutubeParsingHelper.prepareDesktopJsonBuilder;
+import static org.schabi.newpipe.extractor.services.youtube.YoutubeParsingHelper.*;
 import static org.schabi.newpipe.extractor.utils.Utils.UTF_8;
 import static org.schabi.newpipe.extractor.utils.Utils.isNullOrEmpty;
 
@@ -9,6 +8,7 @@ import java.io.IOException;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
+import java.util.OptionalInt;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
@@ -25,6 +25,7 @@ import org.schabi.newpipe.extractor.exceptions.ParsingException;
 import org.schabi.newpipe.extractor.linkhandler.ListLinkHandler;
 import org.schabi.newpipe.extractor.localization.Localization;
 import org.schabi.newpipe.extractor.utils.JsonUtils;
+import org.schabi.newpipe.extractor.utils.Utils;
 
 import com.grack.nanojson.JsonArray;
 import com.grack.nanojson.JsonObject;
@@ -46,6 +47,15 @@ public class YoutubeCommentsExtractor extends CommentsExtractor {
     @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
     private Optional<Boolean> optCommentsDisabled = Optional.empty();
 
+    /**
+     * Caching mechanism and holder of the commentsCount value.
+     * <br/>
+     * Initial value = empty -> unknown if comment a comment count is present or not<br/>
+     * Some method calls {@link #determineCommentsCount(JsonObject)}
+     * -> value is set<br/>
+    * If the method is called again the cached value will be used.
+     */
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private OptionalInt optCommentsCount = OptionalInt.empty();
+
     public YoutubeCommentsExtractor(
             final StreamingService service,
             final ListLinkHandler uiHandler) {
@@ -177,12 +182,36 @@ public class YoutubeCommentsExtractor extends CommentsExtractor {
 
         final JsonObject ajaxJson = getJsonPostResponse("next", body, localization);
 
-        final CommentsInfoItemsCollector collector = new CommentsInfoItemsCollector(
-                getServiceId());
+        final CommentsInfoItemsCollector collector =
+                new CommentsInfoItemsCollector(getServiceId());
+        determineCommentsCount(ajaxJson);
         collectCommentsFrom(collector, ajaxJson);
         return new InfoItemsPage<>(collector, getNextPage(ajaxJson));
     }
 
+    private void determineCommentsCount(final JsonObject ajaxJson) {
+        // Check if commentscount was already determined
+        if (optCommentsCount.isPresent()) {
+            return;
+        }
+
+        final JsonObject countText = ajaxJson
+                .getArray("onResponseReceivedEndpoints").getObject(0)
+                .getObject("reloadContinuationItemsCommand")
+                .getArray("continuationItems").getObject(0)
+                .getObject("commentsHeaderRenderer")
+                .getObject("countText");
+
+        try {
+            this.optCommentsCount = OptionalInt.of(
+                    Integer.parseInt(Utils.removeNonDigitCharacters(getTextFromObject(countText)))
+            );
+        } catch (final Exception e) {
+            // If no count could be determined use the default
+            this.optCommentsCount = OptionalInt.of(NO_COMMENT_COUNT);
+        }
+    }
+
     private void collectCommentsFrom(final CommentsInfoItemsCollector collector,
                                      @Nonnull final JsonObject ajaxJson) throws ParsingException {
 
@@ -257,4 +286,11 @@ public class YoutubeCommentsExtractor extends CommentsExtractor {
 
         return optCommentsDisabled.get();
     }
+
+    @Override
+    public int getCommentsCount() {
+        assertPageFetched();
+
+        return optCommentsCount.orElse(NO_COMMENT_COUNT);
+    }
 }
```

</details>

Reply:
I second @litetex's suggestion, unless I'm missing something

Reply:
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Optional.html

> Optional is primarily intended for use as a method return type where there is a clear need to represent "no result," and where using null is likely to cause errors. A variable whose type is Optional should never itself be null; it should always point to an Optional instance.

I want to remove the optional altogether, it's not meant to be used as a field at all!

Reply:
Yes, I think so too. But then, exception handling will get a little trickier, but that's okay.

Reply:
I've done something different: I've removed the `nextResponse` JSON object and replaced it with `initialToken` which is a String extracted on page fetch.

I really don't think we should be concerned about memory usage since I don't think anyone should be holding extractor objects for long periods. The `CommentInfo` class drops them right after it's used for example, so any performance differences would be _very very_ minimal.

We store response objects anyways in `YoutubePlaylistExtractor`, `YoutubeChannelExtractor`, `YoutubeStreamExtractor`, etc. So, I don't think this is an issue.